### DISCUSSION TITLE
player: key-press overlay

### DISCUSSION THREAD

**Discussion by ku1ik** [asciinema/discussions#154] [Ideas]
New asciicast v2 format has a place for keypresses (https://github.com/asciinema/asciinema/blob/fb1e25d933851743f4aed744279ab5361af9ad4b/doc/asciicast-v2.md#i---data-read-from-stdin) and asciinema 2.0 allows recording stdin (https://github.com/asciinema/asciinema/tree/fb1e25d933851743f4aed744279ab5361af9ad4b#rec-filename). 

The web player could now use recorded stdin events to display these keystrokes in an overlay.

**Comment by shibumi** [comment]
any update on this?

**Comment by ruslanosipov** [comment]
+1

**Comment by sametcodes** [comment]
+1

**Comment by daonb** [comment]
I need this to record sessions where I use both zsh & vim so I want the keys to be displayed only when in vim. I guess I'll have to code it myself, but before I do, I want to make sure I got the design right:

IMHO, there are three components we need:
1. Add support for the [\<track\>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track) sub element to display key presses, captions, etc.  
2. A new utility to convert a recording's stdin to the WebVTT format supported by \<track\>. 
3. in asciinema.org, add the option to upload a WebVTT file when editing a recording 

At first we'll support just a single text track, in the future we can support multiples track without changing the interface.

**Comment by daonb** [comment]
I've started working with creating the subtitle file and it doesn't look good.  Grouping keys together results in captions like "Ihello<esc>" (vim). Problem is this caption will be displayed the second I type `I` and it's too confusing. The keys should be displayed as I type them and that's why I :-1: my idea above.

**Comment by rlanore** [comment]
Any update on this ?

**Comment by bonnefoi** [comment]
Any update on this ?

**Comment by Kreijstal** [comment]
Any update on this?

**Comment by RGD2** [comment]
I'm very interested in any updates on this.

I think this needs higher priority.  

How else to show new users pro tricks like typing SHIFT+(!,$), CTRL+([,E) to insert the last argument from the previous line using readline? Or the way to use proper tab completion?

Ideally, it should be possible in the player to 'step' through each keypress one by one when playing, with only short segments of animation when terminal state changes between keypresses. 

This would make it *far* more useful for what I'd consider asciicinema's 'killer app': **Documentation that can show people *exactly* how to do things**, as if they're watching over your shoulder and can see every keypress. 

Being able to step the replay means they can watch and possibly follow along the sequence at their own pace, which would make it far less confusing / intimidating. 

But they can also watch it at 'full speed', which demonstrates how much time is saved by 'doing it right'.

Else asciicinema replays are more just showing off, and learners will just think *'wow that looks hard, I'll never be able to learn that'*.
Wheras what you want them to come away with is *'wow, that's so fast and easy, how come no-one ever showed me this before?'*.

**Comment by district10** [comment]
Subscribed for years...Modifier keys may can not captured under command line.  Maybe subtitles or https://en.wikipedia.org/wiki/Danmu ?

**Comment by nobodyinperson** [comment]
In the meantime, I made this: https://pypi.org/project/termkeymonitor/

Run in a tmux pane, the pressed keys can be displayed in an asciinema cast.

[![asciicast](https://asciinema.org/a/544131.svg)](https://asciinema.org/a/544131)

**Comment by qaqland** [comment]
Any update on this?

Showing keyboard inputs on website asciinema-player is very helpful!

**Comment by ivan-developer-01** [comment]
Still no updates.

**Comment by ku1ik** [comment]
I agree this would be a valuable addition, and it's something I still plan to implement when I find time. I'd love to spend more time on this project, so please consider becoming a sponsor to make work on asciinema more sustainable - https://github.com/sponsors/ku1ik/. Thanks!

**Reply by mosaleh-dev** [reply]
I am willing to put the time to implement this if provided with guidance

**Comment by ikks** [comment]
Take a look at this PR https://github.com/asciinema/asciinema-player/pull/269
[Screencast from 2024-10-09 20-14-40.webm](https://github.com/user-attachments/assets/7ab4417c-90d2-4207-951c-89eafc0d4d03)

**Comment by eduble** [comment]
Hi,
I took a different approach and wrote a tool called [demoterm](https://github.com/drakkar-lig/demoterm) able to show keypresses on a reserved zone at the bottom of the terminal.
You can combine it with `asciinema` for recording, i.e., `asciinema rec -c demoterm`.

I used it for instance for the terminal recordings of this blog post:
https://walt-project.liglab.fr/blog/2025-11-auto-completion

It works fine on my Ubuntu Linux system (i.e., in `gnome-terminal`). There might be issues if the terminal does not support some of the escape sequences needed (e.g., for saving and restoring cursor position and color modes) though.

**Comment by ku1ik** [comment]
Good news! The newly released [player 3.16.0](https://github.com/asciinema/asciinema-player/releases/tag/v3.16.0) includes keystroke overlay feature.

This is based on @ikks's [initial work](https://github.com/asciinema/asciinema-player/pull/294) (thanks!), with some tweaks. Also thanks to @daonb and @ludolpif for your earlier attempts at this.

It's disabled by default and you can enable it with [keystrokeOverlay](https://docs.asciinema.org/manual/player/options/#keystrokeoverlay) option like this:

```javascript
AsciinemaPlayer.create("/demo.cast", document.getElementById("demo"), {
  keystrokeOverlay: true
});
```

It can also be toggled by pressing the `k` key during playback.

Note that this requires the recording to have [input](https://docs.asciinema.org/manual/asciicast/v3/#i-input-data-read-from-a-terminal) events, as captured by `asciinema rec --capture-input ...`.

The overlay setting is already available on asciinema.org on recording settings page.

Looking forward to hearing about how it works for you. There are a few things that could probably be tweaked (key/fade timeouts, positioning, key presentation etc) but I think it's ok for a first version.

**Comment by nmstoker** [comment]
Great news, can't wait to try it out 🙂
Thank you @ku1ik and all involved

**Comment by ludolpif** [comment]
Very neat !
I am currently quite far from having free time for making sweet usage of 
asciinema, but I hope it will came back later this year.
Thanks for all the fish !
